src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::copy_data() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = m_image_ctx->operations->flatten(m_prog_ctx);
+  if (r < 0) {
+    lderr(m_cct) << "flatten failed: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::disable_mirroring() {
+  if (!m_image_ctx->test_features(RBD_FEATURE_JOURNALING)) {
+    return 0;
+  }
+
+  ldout(m_cct, 10) << dendl;
+
+  int r = cls_client::mirror_image_get(&m_src_io_ctx, m_image_ctx->id,
+                                       &m_src_mirror_image);
+  if (r == -ENOENT) {
+    ldout(m_cct, 10) << "mirroring is not enabled for this image" << dendl;
+    return 0;
+  }
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to retrieve mirror image: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  C_SaferCond ctx;
+  auto req = mirror::DisableRequest<I>::create(m_image_ctx, false, true, &ctx);
+  req->send();
+  r = ctx.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to disable mirroring: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  m_src_migration_spec.mirroring = true;
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::enable_mirroring() {
+  if (!m_image_ctx->test_features(RBD_FEATURE_JOURNALING)) {
+    return 0;
+  }
+
+  bool same_pool = m_src_io_ctx.get_id() == m_dst_io_ctx.get_id();
+
+  if (same_pool &&
+      m_src_mirror_image.state != cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+    return 0;
+  }
+
+  ldout(m_cct, 10) << dendl;
+
+  if (!same_pool) {
+    cls::rbd::MirrorMode mirror_mode;
+    int r = cls_client::mirror_mode_get(&m_dst_io_ctx, &mirror_mode);
+    if (r < 0 && r != -ENOENT) {
+      lderr(m_cct) << "failed to retrieve mirror mode: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+
+    if (mirror_mode == cls::rbd::MIRROR_MODE_DISABLED) {
+      ldout(m_cct, 10) << "mirroring is not enabled for destination pool"
+                       << dendl;
+      return 0;
+    }
+    if (mirror_mode == cls::rbd::MIRROR_MODE_IMAGE &&
+        m_src_mirror_image.state != cls::rbd::MIRROR_IMAGE_STATE_ENABLED) {
+      ldout(m_cct, 10) << "mirroring is not enabled for image" << dendl;
+      return 0;
+    }
+  }
+
+  C_SaferCond ctx;
+  auto req = mirror::EnableRequest<I>::create(m_dst_io_ctx, m_dst_image_id, "",
+                                              m_image_ctx->op_work_queue, &ctx);
+  req->send();
+  int r = ctx.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to enable mirroring: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::reopen_image() {
+  ldout(m_cct, 10) << dendl;
+
+  C_SaferCond on_close;
+  m_image_ctx->state->close(&on_close);
+  int r = on_close.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to close image: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  m_image_ctx->clear_state();
```

Comment:
As an alternate suggestion, if the internal API function were passed a pointer to ```ImageCtx*```, it could close the original and create a new ```ImageCtx``` object w/o the need for the ```clear_state``` function.

Reply:
I think I can avoid `clear_state` here, but I will still need this in future for live migration (when a live client detects "migrating" is set on refresh and reopens the image on the fly). Right now it looks to me I will need to preserve `ImageCtx` object for this to work.

Reply:
@dillaman Combining this with your another suggestion (a new constructor for `RemoveImage`) I tend to agree that creating a new `ImageCtx` instead of reusing the old one may be a better approach (then I would not close the old context and use it for `RemoveImage`).

The problem is though that I can't use the current version of C API function `rbd_migrate(rbd_image_t image, ...)` -- I would need to pass `image` by reference instead, which I don't like much.

On the other hand, we could make migrate API to be similar to `clone`, i.e. pass the source image name, instead of the the context. I made `migrate` to look similar to `copy` API, but actually it might be closer to `clone`.

What do you think?

Reply:
@trociny Yeah -- I like the idea of making it more like clone for simplicity.